### PR TITLE
Add test for np-prompt using pexpect

### PR DESCRIPTION
**Pull request by jseagrave21**
**What current issue(s) does this address, or what feature is it adding?**
Addresses issue #490 
Adds test_prompt.py for testing np-prompt with pexpect

**How did you solve this problem?**
Learned about pexpect, trial and error

**How did you make sure your solution works?**
Using unittest

**Are there any special changes in the code that we should be aware of?**
Yes, requires installation of pexpect

**Please check the following, if applicable:**

- [x] Did you add any tests?
- [x] Did you run `make lint`?
- [x] Did you run `make test`?
- [x] Are you making a PR to a feature branch or development rather than master?
- [x] Did you add an entry to `CHANGELOG.rst`? (if not, please do)

### CODE REVIEW

**Comment by jseagrave21** [comment]
This test meets the original requirements of @metachris 's idea. I would be happy to add more if there is a need. Please review and let me know what you think.

File: requirements.txt [hunk 1]
```diff
@@ -40,6 +40,7 @@ neocore==0.4.10
 numpy==1.14.5
 pbr==4.0.4
 peewee==2.10.2
+pexpect==4.2.1
```
Comment (metachris) [inline requirements.txt]: put this into `requirements_dev.txt` please. the `requirements.txt` file should only contain requirements for running it as user :)

**Comment by metachris** [comment]
Hey, cool start! Quick question -- why so many additions to the CHANGELOG?

File: requirements.txt (hunk 1 above)
Comment (jseagrave21) [reply requirements.txt]: Okay thank you for clarifying.

Comment (jseagrave21) [reply requirements.txt]: Do you mean requirements_doc.txt or should I create a new file? I do not see a requirements_dev.txt

**Comment by jseagrave21** [comment]
I could not figure out how to upload a new file to the development branch when trying to create this PR. (If you look at #490, you can see I already closed on PR). So, this changelog is coming from an older version. I copied it from the newest version and then edited it. It should merge nicely with the current changelog in cityofzion/neo-python/development

**Comment by ixje** [comment]
The test seem to be failing for me
```
(venv) Eriks-Air:neo-python erik$ python -m unittest neo.bin.test_prompt.PromptTest
Enp-prompt running as expected
.
======================================================================
ERROR: test_prompt_open_wallet (neo.bin.test_prompt.PromptTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/erik/Documents/code/neo-python/venv/lib/python3.6/site-packages/pexpect/expect.py", line 108, in expect_loop
    incoming = spawn.read_nonblocking(spawn.maxread, timeout)
  File "/Users/erik/Documents/code/neo-python/venv/lib/python3.6/site-packages/pexpect/pty_spawn.py", line 468, in read_nonblocking
    return super(spawn, self).read_nonblocking(size)
  File "/Users/erik/Documents/code/neo-python/venv/lib/python3.6/site-packages/pexpect/spawnbase.py", line 176, in read_nonblocking
    raise EOF('End Of File (EOF). Empty string style platform.')
pexpect.exceptions.EOF: End Of File (EOF). Empty string style platform.
```

File: requirements_dev.txt [hunk 2]
```diff
@@ -0,0 +1,67 @@
+aenum==2.1.2
```
Comment (metachris) [inline requirements_dev.txt]: sorry, i did not mean to create a `requirements_dev.txt` file. or if you'd do it, then create an empty one with only your new test-related dependencies but not all normal requirements. let's chat in the neopython channel on discord about this.

**Comment by jseagrave21** [comment]
@ixje Can you run this using ..."unittest -v"? I am just trying to see as much as I can. I just ran the test again and get this:

j```
seagrave21@DESKTOP-F7E94IM:~$ cd /mnt/c/users/jseag/neo/neo-python-expect
jseagrave21@DESKTOP-F7E94IM:/mnt/c/users/jseag/neo/neo-python-expect$ source venv/bin/activate
(venv) jseagrave21@DESKTOP-F7E94IM:/mnt/c/users/jseag/neo/neo-python-expect$ alias python=python3.6
(venv) jseagrave21@DESKTOP-F7E94IM:/mnt/c/users/jseag/neo/neo-python-expect$ python -m unittest -v neo/bin/test_prompt.py
test_prompt_open_wallet (neo.bin.test_prompt.PromptTest) ... Opened testwallet.db3 successfully
ok
test_prompt_run (neo.bin.test_prompt.PromptTest) ... np-prompt running as expected
ok

----------------------------------------------------------------------
Ran 2 tests in 21.636s

OK
(venv) jseagrave21@DESKTOP-F7E94IM:/mnt/c/users/jseag/neo/neo-python-expect$
```

File: CHANGELOG.rst [hunk 3]
```diff
@@ -3,17 +3,46 @@ Changelog
...
+- Adds ability to attach a fee to a ``send`` transaction
+- Update Node selection mechanism
+- Store ``Transactions`` list items inside a ``Block`` in a consistent format.
+- Improved peer connection maintenance
+
+
+[0.7.5] 2018-07-19
 -----------------------
-- Updated the requirements
+- Add CreateAddress Feature
```
[the remark this answers is not in the document]
Comment (jseagrave21) [reply CHANGELOG.rst]: done

File: requirements_dev.txt (hunk 2 above)
Comment (jseagrave21) [reply requirements_dev.txt]: okay, I am deleting it

**Comment by metachris** [comment]
Fails for me too :(

    $ python -m unittest -v neo/bin/test_prompt.py
    test_prompt_open_wallet (neo.bin.test_prompt.PromptTest) ... ERROR
    test_prompt_run (neo.bin.test_prompt.PromptTest) ... np-prompt running as expected
    ok

    ======================================================================
    ERROR: test_prompt_open_wallet (neo.bin.test_prompt.PromptTest)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
    File "/Users/chris/Projects/chris/neo/neo-python/venv/lib/python3.6/site-packages/pexpect/expect.py", line 99, in expect_loop
        incoming = spawn.read_nonblocking(spawn.maxread, timeout)
    File "/Users/chris/Projects/chris/neo/neo-python/venv/lib/python3.6/site-packages/pexpect/pty_spawn.py", line 462, in read_nonblocking
        raise TIMEOUT('Timeout exceeded.')
    pexpect.exceptions.TIMEOUT: Timeout exceeded.

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
    File "/Users/chris/Projects/chris/neo/neo-python/neo/bin/test_prompt.py", line 27, in test_prompt_open_wallet
        index = child.expect(['Opened wallet at fixtures/testwallet.db3'])
    File "/Users/chris/Projects/chris/neo/neo-python/venv/lib/python3.6/site-packages/pexpect/spawnbase.py", line 321, in expect
        timeout, searchwindowsize, async)
    File "/Users/chris/Projects/chris/neo/neo-python/venv/lib/python3.6/site-packages/pexpect/spawnbase.py", line 345, in expect_list
        return exp.expect_loop(timeout)
    File "/Users/chris/Projects/chris/neo/neo-python/venv/lib/python3.6/site-packages/pexpect/expect.py", line 107, in expect_loop
        return self.timeout(e)
    File "/Users/chris/Projects/chris/neo/neo-python/venv/lib/python3.6/site-packages/pexpect/expect.py", line 70, in timeout
        raise TIMEOUT(msg)
    pexpect.exceptions.TIMEOUT: Timeout exceeded.
    <pexpect.pty_spawn.spawn object at 0x105164898>
    command: /Users/chris/Projects/chris/neo/neo-python/venv/bin/np-prompt
    args: ['/Users/chris/Projects/chris/neo/neo-python/venv/bin/np-prompt']
    buffer (last 100 chars): b'                                                     \r\x1b[79C \r\x1b[0m\r\r\n\x1b[J\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?2004l'
    before (last 100 chars): b'                                                     \r\x1b[79C \r\x1b[0m\r\r\n\x1b[J\x1b[?7h\x1b[0m\x1b[?12l\x1b[?25h\x1b[?2004l'
    after: <class 'pexpect.exceptions.TIMEOUT'>
    match: None
    match_index: None
    exitstatus: None
    flag_eof: False
    pid: 7790
    child_fd: 5
    closed: False
    timeout: 30
    delimiter: <class 'pexpect.exceptions.EOF'>
    logfile: None
    logfile_read: None
    logfile_send: None
    maxread: 2000
    ignorecase: False
    searchwindowsize: None
    delaybeforesend: 0.05
    delayafterclose: 0.1
    delayafterterminate: 0.1
    searcher: searcher_re:
        0: re.compile("b'Opened wallet at fixtures/testwallet.db3'")

    ----------------------------------------------------------------------
    Ran 2 tests in 31.328s

    FAILED (errors=1)

**Comment by ixje** [comment]
@jseagrave21 correct you cannot approve the changes. The reviewer requests changes, you update and the reviewer again can approve, reject or request more changes.

**Comment by jseagrave21** [comment]
It looks like pexpect isn't seeing the expected text. If you run np-prompt normally and try to open testwallet.db3 like my test shows, do you get the response "Opened wallet at fixtures/testwallet.db3"?

In the meantime, I am uploading an older version my test that was more comprehensive (but I thought might also be redundant). Could you tell me if it works for you?

**Comment by jseagrave21** [comment]
In the meantime, I uploaded an older version of my test that was more comprehensive (but I thought might also be redundant). Could you tell me if it works for you?

**Comment by ixje** [comment]
The new changes make no difference for me. What platform do you use? Me and @metachris both use OSX, so I'm wondering if there's anything going on there.

**Comment by jseagrave21** [comment]
@ixje that may be the source of the problem. I am using Ubuntu 18 on Windows 10.

**Comment by ixje** [comment]
ok, I actually have an update. Changing the expect command to:
```
child = pexpect.spawn('python neo/bin/prompt.py -p')
```
makes it work for me. Now the question is whats different between this and `np-prompt` in terms of execution. I"ll have to investigate tomorrow or friday

**Comment by jseagrave21** [comment]
Okay awesome! I will see if that also works on my end.
It does seem odd to me because spawn is supposed to send a command and search for the command if it isn't immediately found. Do you interact with np-prompt in the same way as described above? https://github.com/CityOfZion/neo-python/pull/536#issuecomment-411500175

**Comment by jseagrave21** [comment]
I am wondering if the problem is that the child application (neo-python) is staying alive after the test is completed. The reason I am wondering this is because the the first test passes for both of you.

Also, the "-p" should mean that it is running on the privatenet right? So the difference is accessing a privatenet vs the testnet?

**Comment by jseagrave21** [comment]
The updated version terminates the child application at the end of each test, and I included a try/except variation, which should allow both tests to pass.

**Comment by metachris** [comment]
Not sure this is still in active development. If it is, could you please clean up the changes to CHANGELOG, so it only has the changes that you actually want? Currently it adds a lot of stuff to the changelog.

Also, another option might be using the normal expect tool rather than the python-based pexpect, and to run it outside our normal test suite, as part of the travis ci. 

I think for running `np-prompt` we should not use `-p` but connect to a normal testnet or mainnet, because that should always work and travis-ci does not run the privnet container anyway.

**Comment by jseagrave21** [comment]
@metachris I have updated the Changelog so it most closely matched the current version in development along with making the edit directed by @ixje

**Comment by jseagrave21** [comment]
@metachris I agree that we should not use the privatenet. The issue we are trying to figure out is why the open_wallet test passes for me running Ubuntu on Windows but does not pass with you and @ixje running OSX.

File: requirements.txt (hunk 1 above)
Comment (metachris) [reply requirements.txt]: just to note -- after internal talks we leave it like this ;)

**Comment by jseagrave21** [comment]
@metachris Is there any reason you think that using expect instead of pexpect will work better in your testing environment?

**Comment by jseagrave21** [comment]
After my success using `subprocess` for np-utils in neo-python-core, I thought I'd try and see if it would work for this purpose but I get this unhandled error:

```
Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/twisted/_threads/_threadworker.py", line 46, in work
    task()
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/twisted/_threads/_team.py", line 190, in doWork
    task()
--- <exception caught here> ---
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/mnt/c/users/jseag/Neo/neo-python-expect/neo/bin/prompt.py", line 937, in run
    print_formatted_text(FormattedText(tokens), style=self.token_style)
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/prompt_toolkit/shortcuts/utils.py", line 91, in print_formatted_text
    output = get_default_output()
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/prompt_toolkit/output/defaults.py", line 61, in get_default_output
    return create_output()
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/prompt_toolkit/output/defaults.py", line 39, in create_output
    stdout, term=get_term_environment_variable())
  File "/mnt/c/users/jseag/Neo/neo-python-expect/venv/lib/python3.6/site-packages/prompt_toolkit/output/vt100.py", line 409, in from_pty
    assert stdout.isatty()
builtins.AssertionError:
```

I am continuing to try to figure out how to make the pexpect test working with OSX and why it may be failing.

**Comment by jseagrave21** [comment]
Okay, I figured out what was happening. I needed to send the `enter` command by using `\n`. @metachris @ixje Please review the updated test.

**Comment by metachris** [comment]
Looks good. Can you please pull the latest development branch, and fix the changes to CHANGELOG (it still shows loads of additions to it that should probably not be part of this PR)?

**Comment by jseagrave21** [comment]
Okay creating a new PR with fresh pull of the development branch.

**Pull request closed without merging**